### PR TITLE
ggml : set output of view src

### PR DESCRIPTION
**Pull request by am17an**
## Overview

The `ggml_set_output` now also sets the src tensors of views as outputs as well. This is to prevent from the allocator from reusing their memory.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure:  no

### CODE REVIEW

**Comment by ggerganov** [comment]
A bit more info would be helpful - where do we use views for output?

**Comment by am17an** [comment]
I'm adding the deepseek4 MTP, there we need to add a hidden state from a reshape (which is a view)

https://github.com/ggml-org/llama.cpp/blob/c3d47e696b1187a27e896aa828d48ff9a33fc679/src/models/deepseek4.cpp#L1156-L1160

**Comment by ggerganov** [comment]
I think you can simply output the hidden state before the reshape? The hidden state extraction logic only needs to know the total size of the tensor - the specific dimensions are not relevant.

**Comment by am17an** [comment]
The unflattened tensor is 3d, `get_embeddings_nextn` expects a 2-d tensor of dim `[n_outputs][n_embd]`

**Comment by am17an** [comment]
Also in general this is a bug right? If any t is a view then this can possibly cause issues. We have no documentation that says that these things can't be views.

**Comment by ggerganov** [comment]
> Also in general this is a bug right? If any t is a view then this can possibly cause issues. We have no documentation that says that these things can't be views.

Yes, we should assert in the `set_outputs()` that none of the outputs are views.

> The unflattened tensor is 3d, `get_embeddings_nextn` expects a 2-d tensor of dim `[n_outputs][n_embd]`

I don't follow. Is this a problem? Normally, all embeddings and hidden states that we extract are 2D.

**Comment by am17an** [comment]
It's not a problem for the MTP, I'll just update this PR to add an assert for outputs not being views then

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
Some server tests are failing after this change: https://github.com/ggml-org/llama.cpp/actions/runs/29490535879/job/87595298014?pr=25729#step:7:752

**Comment by am17an** [comment]
It looks like some cases with backend_sampling: true are failing

**Comment by am17an** [comment]
This is the problem. The top-p default returns a view as output

https://github.com/ggml-org/llama.cpp/blob/fd4fd67b3182fb56bd4feb0ed1997a298de9ce00/src/llama-sampler.cpp#L1434-L1458

**Review by ggerganov — APPROVED** [review]
Should be good to merge. Might have some impact on the https://github.com/ggml-org/llama.cpp/pull/25532

**Comment by gaugarg-nv** [comment]
I faced a similar issue in https://github.com/ggml-org/llama.cpp/pull/25532

How about fixing it within GGML? 

We can propagate the output flag to the backing tensor in `ggml_set_output`. Something like this-

```
void ggml_set_output (struct ggml_tensor * tensor) {
  for (ggml_tensor * cur = tensor; cur != nullptr; cur = cur->view_src) {
      cur->flags |= GGML_TENSOR_FLAG_OUTPUT;
  }
}
```

**Comment by ggerganov** [comment]
I was a bit hesitant to make it since I initially thought that we don't need to use output views. But it does seem that there are some use cases where it's complicated to avoid having a view for output.

**Comment by gaugarg-nv** [comment]
I see that @am17an's first commit in this PR did something similar. But it's not clear to me why an output can't be a "view".
As long as we make sure the backing tensor stays alive during graph execution, we should be fine.

The current assert also enforces this only for tensors registered in `llm_graph_result`. What if `ggml_set_output` is called on a view somewhere else?

**Comment by ggerganov** [comment]
Ok, I added the change to `ggml_set_output`. Will keep the changes to the `llama-sampler.cpp` since I think they improve consistency and should resolve some edge-cases when shaping the input logits of the samplers.

File: src/llama-graph.cpp [hunk 1]
```diff
@@ -3437,6 +3437,7 @@ void llm_graph_context::build_sampling() const {
     // add a dummy row of logits
     // this trick makes the graph static, regardless of which samplers are activated
     // this is important in order to minimize graph reallocations
+    // note: mark this tensor as output since there could be output views of its data
```
Comment (gaugarg-nv) [inline src/llama-graph.cpp]:
```suggestion
```

File: src/llama-sampler.cpp [hunk 2]
```diff
@@ -1435,21 +1440,25 @@ static void llama_sampler_top_p_backend_apply(
...
-    struct ggml_tensor * sorted_idx = ggml_argsort(ctx, data->logits, GGML_SORT_ORDER_DESC);
+    struct ggml_tensor * sorted_idx = ggml_argsort(ctx, logits, GGML_SORT_ORDER_DESC);
     ggml_set_name(sorted_idx, "top_p_sorted_idx");
 
     // Do the sorting via reshape + get_rows
-    struct ggml_tensor * sorted_logits = ggml_sort(data->logits, sorted_idx);
+    struct ggml_tensor * sorted_logits = ggml_sort(logits, sorted_idx);
     ggml_set_name(sorted_logits, "top_p_sorted_logits");
 
+    sorted_logits = ggml_reshape_1d(ctx, sorted_logits, logits->ne[0]);
```
Comment (gaugarg-nv) [inline src/llama-sampler.cpp]:
```suggestion
    sorted_logits = ggml_reshape_1d(ctx, sorted_logits, ggml_nelements(sorted_logits));
```

Nit: IMO, clearer this way.

**Review by gaugarg-nv — APPROVED** [review]
